sdk/python/lib/pulumi/provider/experimental/analyzer.py

```diff
@@ -180,16 +212,55 @@ def analyze_property(
             raise ValueError("list types not yet implemented")
         elif isinstance(arg, dict):
             raise ValueError("dict types not yet implemented")
+        elif is_forward_ref(arg):
+            name = cast(ForwardRef, arg).__forward_arg__
+            type_def = self.type_definitions.get(name)
+            if type_def:
+                # Forward ref to a type we saw before, return a reference to it.
+                ref = f"#/types/{self.metadata.name}:index:{name}"
```

Comment:
Could a user have two types with the same name in different parts of their app? Will they conflict then? It's probably an okay limitation, but ideally we would catch it and fail with a good error message.

Reply:
🤯 yes this could happen for both components and complex types. I added detection here https://github.com/pulumi/pulumi/pull/18419/commits/9f4e961d9c042294927fa98424d109809d94617a by tracking each type's module and raising an error if there are duplicates

```
Duplicate type 'MyComponent': orginally defined in 'lib/test/provider/experimental/testdata/duplicate-type/component_a.py' but also found in 'lib/test/provider/experimental/testdata/duplicate-type/component_b.py'
```